### PR TITLE
API  Implementation

### PR DESCRIPTION
**Pull request by sieradzki**
Implemented basic web API

### CODE REVIEW

**Comment by sieradzki** [comment]
[EEC-03: API Endpoint](https://trello.com/c/1VLjT8YD/17-eec-03-api-endpoint)

File: eec-backend/Controllers/ProductController.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    public class ProductController : ControllerBase
+    {
+        private readonly ProductContext _context;
+
+        public ProductController(ProductContext context)
        {
+            _context = context;
+        }
+
+        // GET: api/Products
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: This type of comments are not really necessary. But to be precise this endpoint is like `api/Product`. You can even check it on swagger :)

File: eec-backend/Controllers/ProductController.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        public async Task<ActionResult<IEnumerable<Product>>> GetProducts()
+        {
+            return await _context.Products.ToListAsync();
+        }
+
+        // GET: api/Products/{modelIdentifier}
+        [HttpGet("{modelIdentifier}")]
+        public async Task<ActionResult<Product>> GetProduct(string modelIdentifier)
+        {
+            var product = await _context.Products.FindAsync(modelIdentifier);
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: we should move this type of logic to Service layer. We dont have any Service classes yet so we need to create one.
Its not .NET specific, its rather general rule in backend to have 3 main layers. Controller, Service and Repository. Please read about it or pm me for more details. Main point is that Controller layer should only be used for communication. All implementation details should be hidden and logic should be in Service.

Comment (Zarpat-hub) [reply eec-backend/Controllers/ProductController.cs]: Repository as name says should be used as layer to get data from database

File: eec-backend/Controllers/ProductController.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            return product;
+        }
+
+        // PUT: api/Products/{modelIdentifier}
+        [HttpPut("{modelIdentifier}")]
+        public async Task<IActionResult> PutProduct(string modelIdentifier, Product product)
+        {
+            if (modelIdentifier != product.ModelIdentifier)
+            {
+                return BadRequest();
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: We should provide a message with details on return.

File: eec-backend/Controllers/ProductController.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            }
+
+            _context.Entry(product).State = EntityState.Modified;
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: we should move this line inside `try catch` block

File: eec-backend/Controllers/ProductController.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                {
+                    return NotFound();
+                }
+                else
+                {
+                    throw;
+                }
+            }
+
+            return NoContent();
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: Here we return when everything goes well so we should rather return `Ok` status (200).

File: eec-backend/Controllers/ProductController.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+            _context.Entry(product).State = EntityState.Modified;
+
+            try
+            {
+                await _context.SaveChangesAsync();
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                if (!ProductExists(modelIdentifier))
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: Please dont throw errors. We should rather return status 500 with exception message to not block the application running. Also move this if statement higher so we can fail faster if no product exists.

Comment (Zarpat-hub) [reply eec-backend/Controllers/ProductController.cs]: `return StatusCode(500, ex.Message)` and add `ex` after `DbUpdateConcurrencyException`

Comment (Zarpat-hub) [reply eec-backend/Controllers/ProductController.cs]: We could also use logger for logging such a details inside our app but this is not really important right now. Maybe I will do it later in some of my PR

File: eec-backend/Controllers/ProductController.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                return NotFound();
+            }
+
+            _context.Products.Remove(product);
+            await _context.SaveChangesAsync();
+
+            return NoContent();
+        }
+
+        private bool ProductExists(string modelIdentifier)
```
Comment (Zarpat-hub) [inline eec-backend/Controllers/ProductController.cs]: Ok we will have only one entity but just to let you know - we can make this method generic and reuse it later with any type of entity. ALso this should not be in controller layer ofc

File: eec-backend/Data/ProductContext.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        public ProductContext(DbContextOptions<ProductContext> options)
+           : base(options)
+        {
+        }
+
+        /* Connection with sql server */
+        //protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        //{
+        //    optionsBuilder.UseSqlServer(@"Data Source=(localdb)\MSSQLLocalDB;Initial Catalog=Api;Integrated Security=True;");
+        //}
```
Comment (Zarpat-hub) [inline eec-backend/Data/ProductContext.cs]: Pm me for connection details to our hosted database. Also we should rather not hardcode it there but anyway in our case. Just get rid of this in memory database in this PR

File: eec-backend/Models/Product.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        [Required]
+        [DatabaseGenerated(DatabaseGeneratedOption.None)]
+        public string? ModelIdentifier { get; set; }
+
+        [Required]
+        public string? SupplierOrTrademark { get; set; }
+
+        [Required]
+        public string? Category { get; set; }
+
```
Comment (Zarpat-hub) [inline eec-backend/Models/Product.cs:23]: You mark this prop as `Required` along with `nullable` modifier. I dont think if it makes sense.

Comment (Zarpat-hub) [reply eec-backend/Models/Product.cs:23]: Also please annotate nullable props with `AllowNull` annotation. Also mark all the props with `JsonProperty(NullValueHandling = NullValueHandling.Ignore)` to not display it on the return if a value is null. Please see attached link for an example
https://www.newtonsoft.com/json/help/html/NullValueHandlingIgnore.htm

File: eec-backend/Program.cs [hunk 10]
```diff
@@ -1,8 +1,16 @@
...
+
 var builder = WebApplication.CreateBuilder(args);
 
 // Add services to the container.
 
 builder.Services.AddControllers();
+
+// For testing purposes
+builder.Services.AddDbContext<ProductContext>(opt =>
+    opt.UseInMemoryDatabase("Eec"));
```
Comment (Zarpat-hub) [inline eec-backend/Program.cs]: Just remember to change it once we got connectionstring

**Review by Zarpat-hub — COMMENTED** [review]
Nice work overall for beginner :D

File: eec-backend/Controllers/ProductController.cs (hunk 2 above)
Comment (Zarpat-hub) [reply eec-backend/Controllers/ProductController.cs]: I can make it in this PR if u dont get it to give you an example

**Comment by Zarpat-hub** [comment]
I have done suggestions from this PR. Take a look, maybe you will find something ;)

**Review by Zarpat-hub — APPROVED** [review]
Merge it, we need to seed database till Friday

**Pull request merged**